### ISSUE TITLE
WGBS error

### ISSUE THREAD

**Issue by pastaxingxin** [maxplanck-ie/snakepipes#702]
My command line: WGBS -i ./wgbs -o ./wgbs-out GRCh38_relase99
There are four data files in ./wgbs directory like this:
![image](https://user-images.githubusercontent.com/17949301/95903746-92758c80-0d64-11eb-918f-28a35b9313c2.png)
However, I got errors after running that command line:
It's something like this:

![image](https://user-images.githubusercontent.com/17949301/95904786-fba9cf80-0d65-11eb-84a7-179b7f390fe6.png)
![image](https://user-images.githubusercontent.com/17949301/95904813-05333780-0d66-11eb-940c-ab82c8695987.png)

How to resolve the problem? Please give me some hints, thank you in advance.

**Comment by LeilyR** [comment]
Did you change the default read names? from the command you have sent it seems not. The default name ending is _R1, _R2. Please either rename your files or change the default using either `--reads` or using a customized config file  and calling it with `-c`.

**Comment by pastaxingxin** [comment]
yes, I changed the default read names like this:
![image](https://user-images.githubusercontent.com/17949301/95984734-a4e9d780-0df1-11eb-9531-1f7b039dbe39.png)

**Comment by pastaxingxin** [comment]
As you could see, 17 of 46 steps (37%) done.

**Comment by LeilyR** [comment]
I see, there was no `-c your_custom.yaml ` in your command line, that is why I have asked.

**Comment by katsikora** [comment]
Hi Pastaxingxin,

do you have a more specific error for this rule in the `cluster_logs` subfolder of your output folder? Or in the `bwameth/logs` subfolder?

Best,

Katarzyna

**Comment by pastaxingxin** [comment]
Thank you for your reply.
![image](https://user-images.githubusercontent.com/17949301/95990612-bf27b380-0df9-11eb-8c10-e13b857e481a.png)
For existing files, only bwameth files shows failed and exited.
bwameth.%j.err is attached here:


---- This analysis has been done using snakePipes version 2.3.0 ----
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 16
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bwameth
	1

[Tue Oct 13 02:55:27 2020]
rule bwameth:
    input: FASTQ_fastp/k562-wgbs-SRR4235743_1.fastq.gz, FASTQ_fastp/k562-wgbs-SRR4235743_2.fastq.gz
    output: bwameth/k562-wgbs-SRR4235743.bam
    log: bwameth/logs/k562-wgbs-SRR4235743.map_reads.err, bwameth/logs/k562-wgbs-SRR4235743.map_reads.out
    jobid: 0
    wildcards: sample=k562-wgbs-SRR4235743
    threads: 16

Activating conda environment: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9
[Tue Oct 13 02:57:16 2020]
Error in rule bwameth:
    jobid: 0
    output: bwameth/k562-wgbs-SRR4235743.bam
    log: bwameth/logs/k562-wgbs-SRR4235743.map_reads.err, bwameth/logs/k562-wgbs-SRR4235743.map_reads.out (check log file(s) for error message)
    conda-env: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9
    shell:
        
            TMPDIR=/data/Frank_Huang_lab/XingxinPan/K562/analysis-chip/data/extended/
            MYTEMP=$(mktemp -d "${TMPDIR:-/tmp}"/snakepipes.XXXXXXXXXX)
            bwameth.py --threads 16 --reference "/data/Frank_Huang_lab/XingxinPan/K562/final/data/huizong/k562/indice/BWAmethIndex/genome.fa" "FASTQ_fastp/k562-wgbs-SRR4235743_1.fastq.gz" "FASTQ_fastp/k562-wgbs-SRR4235743_2.fastq.gz" 2> bwameth/logs/k562-wgbs-SRR4235743.map_reads.err | 	        samtools sort -T "$MYTEMP"/k562-wgbs-SRR4235743 -m 3G -@ 4 -o "bwameth/k562-wgbs-SRR4235743.bam" 2>> bwameth/logs/k562-wgbs-SRR4235743.map_reads.err
            rm -rf "$MYTEMP"
            
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message

---- This analysis has been done using snakePipes version 2.3.0 ----
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 16
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bwameth
	1

[Tue Oct 13 03:22:43 2020]
rule bwameth:
    input: FASTQ_fastp/k562-wgbs-SRR4235744_1.fastq.gz, FASTQ_fastp/k562-wgbs-SRR4235744_2.fastq.gz
    output: bwameth/k562-wgbs-SRR4235744.bam
    log: bwameth/logs/k562-wgbs-SRR4235744.map_reads.err, bwameth/logs/k562-wgbs-SRR4235744.map_reads.out
    jobid: 0
    wildcards: sample=k562-wgbs-SRR4235744
    threads: 16

Activating conda environment: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9
[Tue Oct 13 03:34:22 2020]
Error in rule bwameth:
    jobid: 0
    output: bwameth/k562-wgbs-SRR4235744.bam
    log: bwameth/logs/k562-wgbs-SRR4235744.map_reads.err, bwameth/logs/k562-wgbs-SRR4235744.map_reads.out (check log file(s) for error message)
    conda-env: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9
    shell:
        
            TMPDIR=/data/Frank_Huang_lab/XingxinPan/K562/analysis-chip/data/extended/
            MYTEMP=$(mktemp -d "${TMPDIR:-/tmp}"/snakepipes.XXXXXXXXXX)
            bwameth.py --threads 16 --reference "/data/Frank_Huang_lab/XingxinPan/K562/final/data/huizong/k562/indice/BWAmethIndex/genome.fa" "FASTQ_fastp/k562-wgbs-SRR4235744_1.fastq.gz" "FASTQ_fastp/k562-wgbs-SRR4235744_2.fastq.gz" 2> bwameth/logs/k562-wgbs-SRR4235744.map_reads.err | 	        samtools sort -T "$MYTEMP"/k562-wgbs-SRR4235744 -m 3G -@ 4 -o "bwameth/k562-wgbs-SRR4235744.bam" 2>> bwameth/logs/k562-wgbs-SRR4235744.map_reads.err
            rm -rf "$MYTEMP"
            
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message

**Comment by pastaxingxin** [comment]
For your information, I have run createIndices and DNA-mapping successfully, but got error messages in WGBS, CHiP-seq, mRNA-seq.

**Comment by katsikora** [comment]
Hi,

so on our side, the WGBS pipeline runs successfully with our test data if I use _1,_2 as reads values.

These error logs from `cluster_logs` folder are uninformative.
Although this is probably unrelated, the `%j` in your snakemake_cluster_cmd doesn't evaluate to job number on your cluster. I guess you are running on something else than a slurm cluster?

In the `bwameth` folder, there is a `logs` subfolder with bwameth logs in it. Does this have any relevant information?

It's difficult to say why a number of workflows are failing. Would you like to try with our test data?

See https://snakepipes.readthedocs.io/en/latest/content/setting_up.html bottom of the page for the list of test datasets we provide on zenodo.

Best,

Katarzyna

**Comment by pastaxingxin** [comment]
Thank you for your reply. Yes. It's an LSF cluster. 
Yes, bwameth logs have some relevant information. You can see here:

running: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/python /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py c2t FASTQ_f
converting reads in FASTQ_fastp/k562-wgbs-SRR4235744_1.fastq.gz,FASTQ_fastp/k562-wgbs-SRR4235744_2.fastq.gz
[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 1631290 sequences (160000120 bp)...
[M::process] 0 single-end sequences; 1631290 paired-end sequences
[M::process] read 1630690 sequences (160000010 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 583596, 43, 0)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (159, 193, 208)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (61, 306)
[M::mem_pestat] mean and std.dev: (182.67, 43.03)
[M::mem_pestat] low and high boundaries for proper pairs: (11, 355)
[M::mem_pestat] analyzing insert size distribution for orientation RF...
[M::mem_pestat] (25, 50, 75) percentile: (163, 173, 201)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (87, 277)
[M::mem_pestat] mean and std.dev: (170.91, 22.86)
[M::mem_pestat] low and high boundaries for proper pairs: (49, 315)
[M::mem_pestat] skip orientation RR as there are not enough pairs
[M::mem_pestat] skip orientation RF
[M::mem_process_seqs] Processed 1631290 reads in 2078.928 CPU sec, 656.302 real sec
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 509, in <module>
    main(sys.argv[1:])
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 502, in main
    bwa_mem(args.reference, conv_fqs_cmd, ' '.join(map(str, pass_through_args)),
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 331, in bwa_mem
    as_bam(cmd, fa, set_as_failed)
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 354, in as_bam
    sys.stdout.write(str(aln) + '\n')
BrokenPipeError: [Errno 32] Broken pipe
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 509, in <module>
    main(sys.argv[1:])
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 473, in main
    sys.exit(convert_reads(args[1], args[2]))
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 133, in convert_reads
    convert_and_write_read(name,seq,qual,read_i%2,out)
  File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 166, in convert_and_write_read
    out.write("".join((name, seq, "\n+\n", qual)))
BrokenPipeError: [Errno 32] Broken pipe
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe

I will try with your test data and tell you results soon.

**Comment by pastaxingxin** [comment]
It still cannot work well. Could you please give me some suggestions about the err messages here?

running: /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/python /usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py c2t FASTQ_f
converting reads in FASTQ_fastp/k562-wgbs-SRR4235744_1.fastq.gz,FASTQ_fastp/k562-wgbs-SRR4235744_2.fastq.gz
[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 1631290 sequences (160000120 bp)...
[M::process] 0 single-end sequences; 1631290 paired-end sequences
[M::process] read 1630690 sequences (160000010 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 583596, 43, 0)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (159, 193, 208)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (61, 306)
[M::mem_pestat] mean and std.dev: (182.67, 43.03)
[M::mem_pestat] low and high boundaries for proper pairs: (11, 355)
[M::mem_pestat] analyzing insert size distribution for orientation RF...
[M::mem_pestat] (25, 50, 75) percentile: (163, 173, 201)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (87, 277)
[M::mem_pestat] mean and std.dev: (170.91, 22.86)
[M::mem_pestat] low and high boundaries for proper pairs: (49, 315)
[M::mem_pestat] skip orientation RR as there are not enough pairs
[M::mem_pestat] skip orientation RF
[M::mem_process_seqs] Processed 1631290 reads in 2078.928 CPU sec, 656.302 real sec
Traceback (most recent call last):
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 509, in
main(sys.argv[1:])
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 502, in main
bwa_mem(args.reference, conv_fqs_cmd, ' '.join(map(str, pass_through_args)),
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 331, in bwa_mem
as_bam(cmd, fa, set_as_failed)
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 354, in as_bam
sys.stdout.write(str(aln) + '\n')
BrokenPipeError: [Errno 32] Broken pipe
Traceback (most recent call last):
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 509, in
main(sys.argv[1:])
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 473, in main
sys.exit(convert_reads(args[1], args[2]))
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 133, in convert_reads
convert_and_write_read(name,seq,qual,read_i%2,out)
File "/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9/bin/bwameth.py", line 166, in convert_and_write_read
out.write("".join((name, seq, "\n+\n", qual)))
BrokenPipeError: [Errno 32] Broken pipe
Exception ignored in: <_io.TextIOWrapper name='' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe

Thank you in advance.

**Comment by katsikora** [comment]
The error log you've pasted above is not from our test data, right? We don't have SRR4235744 in there.

It looks like bwa-meth cannot pipe the output to samtools. 
This can have multiple reasons e.g. if your $TMPDIR is full or you have no permissions to write there.
Or samtools is failing for some other reason.

If you source activate your conda env `/usr/local/anaconda3/envs/c29a3c46aa262224b7224ddcc93211c9`, do you get output for `samtools --help` ?

You can also debug the full command by source activating your env, and then pasting shell commands one by one , perhaps removing the stderr redirection: 
TMPDIR=/data/Frank_Huang_lab/XingxinPan/K562/analysis-chip/data/extended/
        MYTEMP=$(mktemp -d "${TMPDIR:-/tmp}"/snakepipes.XXXXXXXXXX)
        bwameth.py --threads 16 --reference "/data/Frank_Huang_lab/XingxinPan/K562/final/data/huizong/k562/indice/BWAmethIndex/genome.fa" "FASTQ_fastp/k562-wgbs-SRR4235744_1.fastq.gz" "FASTQ_fastp/k562-wgbs-SRR4235744_2.fastq.gz"  | 	  samtools sort -T "$MYTEMP"/k562-wgbs-SRR4235744 -m 3G -@ 4 -o "bwameth/k562-wgbs-SRR4235744.bam" 
        rm -rf "$MYTEMP"

Maybe this will catch the samtools error.

**Comment by pastaxingxin** [comment]
Thank you very much. I will check it and see how it is going.

**Comment by pastaxingxin** [comment]
Hi,
I found my samtools couldn't work in this conda environment, which is 1.7 version. So I upgrade it to 1.10, however, a new error message arose:
/usr/bin/bash: line 2: bwameth.py: command not found
samtools sort: failed to read header from "-"

Could you give me some suggestions for this issue? 
Thank you very much.

**Comment by katsikora** [comment]
Hi ,

the error says you're missing bwameth in your conda env.

Best,

Katarzyna

**Comment by alexgreenlon** [comment]
@pastaxingxin I was getting the same error that you initially posted and thanks to your follow-up comment found that `snakePipes createEnvs` installs a broken version of samtools in all of the conda environments it creates (this was true both on an AWS instance running ubuntu and on my macbook). My solution was to just replace the samtools binary in all the snakePipes conda envs with a symlink to a local installation of samtools I compiled from source.

Something like this (though this will also put the same symlink in any other conda environments you have, so maybe specify a list of the conda envs you want to fix):
```
for line in `ls -1 /data1/conda/envs/`; do
rm /data1/conda/envs/"$line"/bin/samtools
ln -s /data1/samtools-1.11/samtools /data1/conda/envs/"$line"/bin/samtools; done
```

**Comment by katsikora** [comment]
Hi Alexgreenlon,

thanks for pinpointing the issue. This has not been the case on our system (CentOS7).
We can rebuild the package with a different samtools version next time.

Maybe there is some temporary issue with samtools  1.7 binary on conda.

**Comment by katsikora** [comment]
The fix is now available in release 2.4.2.
